sparrow-py/pysrc/sparrow_py/expr.py

```diff
@@ -0,0 +1,209 @@
+        ValueError
+            When using `self` with a specific `keyword` if the `keyword` also
+            appears on in the `kwargs`.
+
+        Notes
+        -----
+        Use ``.pipe`` when chaining together functions that expect
+        expressions.
+        Instead of writing
+
+        >>> func(g(h(df), arg1=a), arg2=b, arg3=c)  # doctest: +SKIP
+
+        You can write
+
+        >>> (df.pipe(h)
+        ...    .pipe(g, arg1=a)
+        ...    .pipe(func, arg2=b, arg3=c)
+        ... )  # doctest: +SKIP
+
+        If you have a function that takes the data as (say) the second
+        argument, pass a tuple indicating which keyword expects the
+        data. For example, suppose ``func`` takes its data as ``arg2``:
+
+        >>> (df.pipe(h)
+        ...    .pipe(g, arg1=a)
+        ...    .pipe((func, 'arg2'), arg1=a, arg3=c)
+        ...  )  # doctest: +SKIP
+        """
+        if isinstance(func, tuple):
+            func, target = func
+            if target in kwargs:
+                msg = f"{target} is both the pipe target and a keyword argument"
+                raise ValueError(msg)
+            kwargs[target] = self
+            return func(*args, **kwargs)
+        else:
+            return func(self, *args, **kwargs)
+
+    def __getattr__(self, name: str) -> "Expr":
+        """
+        Create an expression referencing the given field.
+
+        Parameters
+        ----------
+        name : str
+            Name of the field to access.
+
+        Returns
+        -------
+        Expr
+            Expression referencing the given field.
+        """
+        return Expr("fieldref", self, Literal(name))
+
+    def __add__(self, rhs: "Expr") -> "Expr":
+        """Add two expressions."""
+        return Expr("add", self, rhs)
+
+    def __sub__(self, rhs: "Expr") -> "Expr":
+        """Subtract two expressions."""
+        return Expr("sub", self, rhs)
+
+    def __mul__(self, rhs: "Expr") -> "Expr":
+        """Multiple two expressions."""
+        return Expr("mul", self, rhs)
+
+    def __truediv__(self, rhs: "Expr") -> "Expr":
+        """Divide two expressions."""
+        return Expr("div", self, rhs)
+
+    def __lt__(self, rhs: "Expr") -> "Expr":
+        """Less than comparison."""
+        return Expr("lt", self, rhs)
+
+    def __le__(self, rhs: "Expr") -> "Expr":
+        """Less than or equal comparison."""
+        return Expr("le", self, rhs)
+
+    def __gt__(self, rhs: "Expr") -> "Expr":
+        """Greater than comparison."""
+        return Expr("gt", self, rhs)
+
+    def __ge__(self, rhs: "Expr") -> "Expr":
+        """Greater than or equal comparison."""
+        return Expr("ge", self, rhs)
+
+    def __and__(self, rhs: "Expr") -> "Expr":
+        """Logical and."""
+        return Expr("and", self, rhs)
+
+    def __or__(self, rhs: "Expr") -> "Expr":
+        """Logical or."""
+        return Expr("or", self, rhs)
+
+    def __getitem__(self, key: Union["Expr", int, float, str]) -> "Expr":
```

Comment:
This allows:

`expr['foo']` on structs, and `expr['bar']` on maps, etc.

One thing to note here is that it uses `Union["Expr", int, float, str]` instead of the more typical `Expr`. This allows passing a literal in without explicitly annotating it (eg., `Literal(5)` makes an expressions, but you can omit it here).

My general feeling is that *in general* we should prefer `Expr`, since that requires the `Literal(5)` and makes it clear that we're "freezing" that value. Otherwise, we could have:

`five + date.now()` and that would be treated as `five + Literal(date.now())` -- it would freeze the `date.now()` at whatever Python returned when the query was created, rather than the re-evaluating it each time. It seems like in many places the explicit `Literal(date.now())` to change a python value into an expression is desirable.